llvm/lib/Target/PowerPC/PPCISelLowering.cpp

```diff
@@ -3412,13 +3416,23 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddressAIX(SDValue Op,
     return DAG.getNode(PPCISD::ADD_TLS, dl, PtrVT, TLSReg, VariableOffset);
   }
 
-  // Only Local-Exec, Initial-Exec and General-Dynamic TLS models are currently
-  // supported models. If Local- or Initial-exec are not possible or specified,
-  // all GlobalTLSAddress nodes are lowered using the general-dynamic model.
-  // We need to generate two TOC entries, one for the variable offset, one for
-  // the region handle. The global address for the TOC entry of the region
-  // handle is created with the MO_TLSGDM_FLAG flag and the global address
-  // for the TOC entry of the variable offset is created with MO_TLSGD_FLAG.
+  if (Model == TLSModel::LocalDynamic) {
+    // For local-dynamic on AIX, we need to generate two TOC entries, one for
+    // the variable offset, the other for the module handle. The module handle
+    // is encapsulated inside the TLSLD_AIX pseudo node, and will be expanded by
+    // PPCTLSDynamicCall.
+    SDValue VariableOffsetTGA =
+        DAG.getTargetGlobalAddress(GV, dl, PtrVT, 0, PPCII::MO_TLSLD_FLAG);
+    SDValue VariableOffset = getTOCEntry(DAG, dl, VariableOffsetTGA);
```

Comment:
Is it missing TOC entry for the module handle?

Reply:
+1. We can add another flag `PPCII::MO_TLSLDM_FLAG`(Maybe there is not enough bit in `PPCII`, we can use a distinct value, or extend the MO_ACCESS_MASK to free more bits.).

With this way, seems we don't need the "strange" function `isSpecialAIXSymbolTLSML()` in the PPCAsmPrinter file.

Reply:
I agree, I was thinking we need the TOC entry for the module handle.

Reply:
There is a blocker which limits FLAG bit size to be 12:

```
llvm/include/llvm/CodeGen/MachineOperand.h
...
  unsigned SubReg_TargetFlags : 12;
...
  void setTargetFlags(unsigned F) {
    assert(!isReg() && "Register operands can't have target flags");
    SubReg_TargetFlags = F;
    assert(SubReg_TargetFlags == F && "Target flags out of range");
  } 
```

Currently PPCII::TOF is already full. I'm afraid this will be a FIXME...

Reply:
I still didn't figure out how to remove the strange function.

The way I tried was getOrCreate ExternalLinkage GV named "_$TLSML", and call getTOCEntry on that GV. However there will be a symbol table entry for the symbol, but we don't want that.

For asm mode, I can add hack in MCAsmStreamer::emitXCOFFSymbolLinkageWithVisibility to do not emit the ".extern _Renamed..5f24__TLSML; rename _Renamed..5f24__TLSML ......".

For obj mode, things are more complex, and I'm not able to cleanly remove the symbol table entry without by using hack approach.

Reply:
I don't think the module handle has external linkage, `_$TLSML` is renamed from the module handle defined in the module. If we need a `GlobalValue` to create a TOC entry for the module handle, `PreISelIntrinsicLowering` would be a candidate pass to realize it.

Reply:
So I did have another version which can essentially remove the assert check with some side effects:

(1) It did generated extra external symbol reference, which looks like no harm with simple HelloWorld. (Notice: xlC does not generate this symbol reference.)
```
        .extern _Renamed..5f24__TLSML[UA]
        .rename _Renamed..5f24__TLSML[UA],"_$TLSML"
```
(2) Inside `XCOFFObjectWriter::recordRelocation()` logic I have to apply a hack to get the correct symbol, because now there are two symbols: one is `_Renamed..5f24__TLSML[UA]`, the other is `_Renamed..5f24__TLSML[TC]`. Without hack it will choose the [UA] one which will cause linker failure.

Attached the draft patch: [ReplaceAssertHack.diff.txt](https://github.com/llvm/llvm-project/files/12723368/ReplaceAssertHack.diff.txt)

I think current version has the advantage that it does not generate any extra external symbol reference or any extra relocation entry, although it looks ugly with regard to those asserts.

Is it possible that we fix the hack implementation with some future version (I will open an issue and work on it afterwards)? How about let's move on with current approach? 

@amy-kwan @stephenpeckham @bzEq @chenzheng1030 Appreciate your comments. Thank you!

Reply:
4b932d84f48e0f3f42c769a5ca7ce6623ab62f2e is committed to redesign the target flags of machine operand on PPC. Now we should be able to add new flags.